### PR TITLE
feat(compaction): support serverless compaction

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## Checklist

~~- [ ] I have written necessary rustdoc comments~~
~~- [ ] I have added necessary unit tests and integration tests~~
~~- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~~
~~- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
~~- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)~~

~~- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))~~

## Documentation

~~- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)~~

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/compactor/src/server.rs [hunk 1]
```diff
@@ -272,3 +279,261 @@ pub async fn compactor_serve(
...
+    let config = load_config(&opts.config_path, &opts);
+    info!("Starting compactor node",);
+    info!("> config: {:?}", config);
+    info!(
+        "> debug assertions: {}",
+        if cfg!(debug_assertions) { "on" } else { "off" }
+    );
+    info!("> version: {} ({})", RW_VERSION, GIT_SHA);
+
+    // In dedicated compaction mode, these parameters are load from storage opt,
```
Comment (wcy-fdu) [inline src/storage/compactor/src/server.rs:299]: These k8s pod level parameters should be passed by proxy via cloud means.

File: src/storage/compactor/src/server.rs [hunk 2]
```diff
@@ -272,3 +279,261 @@ pub async fn compactor_serve(
...
+    let block_size_kb: u32 = 0;
+    let object_store_recv_buffer_size: usize = 0;
+    let sstable_size_mb: u32 = 0;
+    let bloom_false_positive: f64 = 0.0;
+    let compactor_max_sst_size: u64 = 0;
+    let compact_iter_recreate_timeout_ms: u64 = 0;
+
+    let meta_cache_capacity_mb: usize = 0;
+
+    // in shared compaction mode, these object stoage related parameters should be defined via cloud
```
Comment (wcy-fdu) [inline src/storage/compactor/src/server.rs:313]: These object storage level parameters should be passed by proxy via cloud means.

File: src/storage/compactor/src/server.rs [hunk 3]
```diff
@@ -272,3 +279,261 @@ pub async fn compactor_serve(
...
 }
+
+pub async fn shared_compactor_serve(
+    listen_addr: SocketAddr,
+    advertise_addr: HostAddr,
+    opts: CompactorOpts,
+) -> (JoinHandle<()>, Sender<()>) {
+    type CompactorMemoryCollector = HummockMemoryCollector;
+
+    let config = load_config(&opts.config_path, &opts);
```
Comment (wcy-fdu) [inline src/storage/compactor/src/server.rs:290]: Not sure how k8s pod gets RWConfig.

File: src/storage/compactor/src/server.rs [hunk 4]
```diff
@@ -272,3 +279,261 @@ pub async fn compactor_serve(
...
+        AsyncStackTraceOption::Off => None,
+        c => await_tree::ConfigBuilder::default()
+            .verbose(c.is_verbose().unwrap())
+            .build()
+            .ok(),
+    };
+    let await_tree_reg =
+        await_tree_config.map(|c| Arc::new(RwLock::new(await_tree::Registry::new(c))));
+
+    // The following will be passed via DispatchCompactionTaskRequest, so here is just a simulation.
```
Comment (wcy-fdu) [inline src/storage/compactor/src/server.rs:435]: WIP,  `output_ids`, `id_to_table`, `dispatch_task` will be changed to receive RPC to get.

File: src/storage/src/hummock/compactor/mod.rs [hunk 5]
```diff
@@ -593,3 +623,160 @@ pub fn start_compactor(compactor_context: Arc<CompactorContext>) -> (JoinHandle<
...
+            dispatch_compaction_task_request::Task::VacuumTask(vacuum_task) => {
+                match Vacuum::handle_vacuum_task(
+                    sstable_store.clone(),
+                    &vacuum_task.sstable_object_ids,
+                )
+                .await
+                {
+                    Ok(_) => {
+                        // todo(wcy): report VacuumTask via ReportCompactionTaskRequest rpc.
+                        let report_compaction_task_request = ReportCompactionTaskRequest {
```
Comment (wcy-fdu) [inline src/storage/src/hummock/compactor/mod.rs:765]: WIP, will send `report_compaction_task_request ` to proxy via ReportCompactionTaskRequest.

File: src/storage/src/hummock/compactor/mod.rs [hunk 6]
```diff
@@ -593,3 +623,160 @@ pub fn start_compactor(compactor_context: Arc<CompactorContext>) -> (JoinHandle<
...
+                    bloom_false_positive,
+                    compactor_max_sst_size,
+                    compact_iter_recreate_timeout_ms,
+                    await_tree_reg,
+                )
+                .await;
+                shutdown.lock().unwrap().remove(&task_id);
+                running_task_count.fetch_sub(1, Ordering::SeqCst);
+                // todo: compactor pod send CompactTask via ReportCompactionTaskRequest
+                let report_compaction_task_request = ReportCompactionTaskRequest {
```
Comment (wcy-fdu) [inline src/storage/src/hummock/compactor/mod.rs:749]: WIP, will send `ReportTask`  to proxy via ReportCompactionTaskRequest.

File: src/storage/src/hummock/compactor/mod.rs [hunk 7]
```diff
@@ -593,3 +623,160 @@ pub fn start_compactor(compactor_context: Arc<CompactorContext>) -> (JoinHandle<
...
+                        task_id,
+                        num_ssts_sealed: progress.num_ssts_sealed.load(Ordering::Relaxed),
+                        num_ssts_uploaded: progress.num_ssts_uploaded.load(Ordering::Relaxed),
+                        num_progress_key: progress.num_progress_key.load(Ordering::Relaxed),
+                        num_pending_read_io: progress.num_pending_read_io.load(Ordering::Relaxed) as u64,
+                        num_pending_write_io: progress.num_pending_write_io.load(Ordering::Relaxed) as u64,
+                    });
+                }
+
+                let report_compaction_task_request = ReportCompactionTaskRequest{
```
Comment (wcy-fdu) [inline src/storage/src/hummock/compactor/mod.rs:678]: WIP, will send `HeartBeat` to proxy via ReportCompactionTaskRequest.

File: src/storage/compactor/src/server.rs [hunk 8]
```diff
@@ -272,3 +284,192 @@ pub async fn compactor_serve(
 
     (join_handle, observer_join_handle, shutdown_send)
 }
+
+pub async fn shared_compactor_serve(
```
Comment (wcy-fdu) [inline src/storage/compactor/src/server.rs:288]: entrance of serverless compaction

File: src/storage/src/hummock/compactor/mod.rs [hunk 9]
```diff
@@ -593,3 +626,176 @@ pub fn start_compactor(compactor_context: Arc<CompactorContext>) -> (JoinHandle<
...
+                        compact_task: Some(compact_task),
+                        table_stats_change: to_prost_table_stats_map(table_stats),
+                    })),
+                };
+
+                // match client.repory_compaction_task(report_compaction_task_request).await{
+                //     Ok(_) => {},
+                //     Err(e) => tracing::warn!("Failed to report task {task_id:?} . {e:?}"),
+                // }
+            }
```
Comment (wcy-fdu) [inline src/storage/src/hummock/compactor/mod.rs:733]: Currently there is no report_compaction_task interface in `HummockManagerService`.

File: src/storage/compactor/src/lib.rs [hunk 10]
```diff
@@ -89,6 +89,15 @@ pub struct CompactorOpts {
...
     pub object_store_read_timeout_ms: Option<u64>,
+
+    pub is_shared_compactor: bool,
+    endpoint: String,
+    parallel_compact_size_mb: u32,
+    sstable_size_mb: u32,
+    block_size_kb: u32,
+    bloom_false_positive: f64,
+    state_store_url: String,
+    data_directory: String,
```
Comment (hzxa21) [inline src/storage/compactor/src/lib.rs:100]: You need to add the `#[clap(...)]` before each opts. Otherwise they won't show up in the startup opts.

File: src/storage/compactor/src/lib.rs [hunk 11]
```diff
@@ -89,6 +89,15 @@ pub struct CompactorOpts {
     #[clap(long, env = "RW_OBJECT_STORE_READ_TIMEOUT_MS", value_enum)]
     #[override_opts(path = storage.object_store_read_timeout_ms)]
     pub object_store_read_timeout_ms: Option<u64>,
+
+    pub is_shared_compactor: bool,
+    endpoint: String,
```
Comment (hzxa21) [inline src/storage/compactor/src/lib.rs:94]: Need a better naming to indicate what endpoint this is. How about just naming it `proxy_endpoint`?

File: src/storage/compactor/src/lib.rs [hunk 12]
```diff
@@ -89,6 +89,15 @@ pub struct CompactorOpts {
...
+    parallel_compact_size_mb: u32,
+    sstable_size_mb: u32,
+    block_size_kb: u32,
+    bloom_false_positive: f64,
```
Comment (hzxa21) [inline src/storage/compactor/src/lib.rs:98]: Do we need to provide this global settings to the compactor? I thought it is controlled by compaction group configs and are set per task. cc @Li0k

File: src/storage/compactor/src/lib.rs (hunk 11 above)
Comment (wcy-fdu) [reply src/storage/compactor/src/lib.rs:94]: Agree, let's call it `proxy_rpc_endpoint`

File: src/storage/src/hummock/compactor/mod.rs [hunk 13]
```diff
@@ -631,3 +643,199 @@ pub fn start_compactor(
...
+    let task_progress = context.task_progress_manager.clone();
+    let (shutdown_tx, mut shutdown_rx) = tokio::sync::oneshot::channel();
+    let periodic_event_update_interval = Duration::from_millis(1000);
+
+    let join_handle = tokio::spawn(async move {
+        let shutdown_map = CompactionShutdownMap::default();
+        let report_client = client.clone();
+        let mut periodic_event_interval = tokio::time::interval(periodic_event_update_interval);
+        let executor = context.compaction_executor.clone();
+        let mut heart_beat_client = client.clone();
```
Comment (Li0k) [inline src/storage/src/hummock/compactor/mod.rs:666]: Can we clone the client only once per task?

File: src/storage/src/hummock/compactor/mod.rs [hunk 14]
```diff
@@ -631,3 +643,199 @@ pub fn start_compactor(
...
+
+        'consume_stream: loop {
+            let request: Option<Request<DispatchCompactionTaskRequest>> = tokio::select! {
+                _ = periodic_event_interval.tick() => {
+                    let mut progress_list = Vec::new();
+                    for (&task_id, progress) in task_progress.lock().iter() {
```
Comment (Li0k) [inline src/storage/src/hummock/compactor/mod.rs:672]: Encapsulating CompactTaskProgress logic.

File: src/storage/src/hummock/compactor/mod.rs [hunk 15]
```diff
@@ -631,3 +643,199 @@ pub fn start_compactor(
...
+                                            table_stats_change: to_prost_table_stats_map(table_stats),
+                                        })),
+                                    };
+
+                                    match report_client
+                                        .report_compaction_task(report_compaction_task_request)
+                                        .await
+                                    {
+                                        Ok(_) => {}
+                                        Err(e) => tracing::warn!("Failed to report task {task_id:?} . {e:?}"),
```
Comment (Li0k) [inline src/storage/src/hummock/compactor/mod.rs:764]: TODO: Consider analyzing this err to decide if we should retry or cancel the task

File: src/storage/compactor/src/server.rs [hunk 16]
```diff
@@ -48,23 +49,106 @@ use risingwave_storage::monitor::{
...
 use tracing::info;
 
 use super::compactor_observer::observer_manager::CompactorObserverNode;
 use crate::rpc::{CompactorServiceImpl, MonitorServiceImpl};
 use crate::telemetry::CompactorTelemetryCreator;
 use crate::CompactorOpts;
 
+pub async fn prepare_start_parameters(
+    config: RwConfig,
+
```
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/storage/compactor/src/server.rs:65]: WIP, will merge more redundant code later.

File: src/meta/src/rpc/service/hummock_service.rs [hunk 17]
```diff
@@ -523,4 +524,30 @@ where
 
         Ok(Response::new(RwReceiverStream::new(rx)))
     }
+
+    async fn report_compaction_task(
```
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/meta/src/rpc/service/hummock_service.rs:528]: We put two types in ReportCompactionTaskRequest, ReportTask or HeartBeat.

File: src/storage/hummock_test/src/compactor_tests.rs [hunk 18]
```diff
@@ -233,11 +218,10 @@ pub(crate) mod tests {
             ) => rpc_filter_key_extractor_manager,
             FilterKeyExtractorManager::StaticFilterKeyExtractorManager(_) => unreachable!(),
```
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/storage/hummock_test/src/compactor_tests.rs:219]: This is only used in unit tests.

File: src/storage/compactor/src/lib.rs (hunk 12 above)
Comment (wcy-fdu) [reply src/storage/compactor/src/lib.rs]: After some offline discussion, we decide to reuse system params, and to not put these global settings in `CompactorOpts`, every time starting a new compactor pod, it will get system params via `SystemParamsServiceClient`.

File: src/connector/src/parser/protobuf/recursive.rs [hunk 19]
```diff
@@ -0,0 +1,157 @@
+#[allow(clippy::derive_partial_eq_without_eq)]
```
Comment (hzxa21) [inline src/connector/src/parser/protobuf/recursive.rs:1]: This file seems unrelated. Is there something wrong with the rebase/merge?

Comment (wcy-fdu) [reply src/connector/src/parser/protobuf/recursive.rs:1]: Wired, let me remove it.

File: src/meta/src/hummock/compaction/mod.rs [hunk 20]
```diff
@@ -283,12 +283,6 @@ pub fn create_compaction_task(
 ) -> CompactionTask {
     let target_file_size = if input.target_level == 0 {
         compaction_config.target_file_size_base
-    } else if input.target_level == base_level {
-        // This is just a temporary optimization measure. We hope to reduce the size of SST as much
-        // as possible to reduce the amount of data blocked by a single task during compaction,
-        // but too many files will increase computing overhead.
-        // TODO: remove it after can reduce configuration `target_file_size_base`.
-        compaction_config.target_file_size_base / 4
```
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/meta/src/hummock/compaction/mod.rs:291]: Will we fix this in another PR? IIRC, it's a bug.

File: src/storage/Cargo.toml [hunk 21]
```diff
@@ -67,6 +67,7 @@ tokio = { version = "0.2", package = "madsim-tokio", features = [
     "signal",
 ] }
 tokio-retry = "0.3"
+tonic = { workspace = true }
```
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/storage/Cargo.toml:70]: We will use tonic(`tonic::transport::Channel;`) in storage folder.

File: src/storage/compactor/src/lib.rs [hunk 22]
```diff
@@ -97,28 +103,46 @@ use std::pin::Pin;
...
-        let (join_handle, observer_join_handle, _shutdown_sender) =
-            compactor_serve(listen_addr, advertise_addr, opts).await;
-
-        join_handle.await.unwrap();
-        observer_join_handle.abort();
-    })
+    match opts.compactor_mode.as_str() {
+        "shared" => Box::pin(async move {
+            tracing::info!("Shared compactor pod options: {:?}", opts);
+            tracing::info!("meta address: {}", opts.meta_address.clone());
```
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/storage/compactor/src/lib.rs:109]: It seems that the final implementation is that the proxy does not pretend to be meta, so the log endpoint is more reasonable.

File: src/storage/src/hummock/compactor/mod.rs [hunk 23]
```diff
@@ -374,7 +382,7 @@ pub fn start_compactor(
 
             let executor = compactor_context.compaction_executor.clone();
             let sstable_object_id_manager = sstable_object_id_manager.clone();
-            let mut last_workload = CompactorWorkload::default();
+            let last_workload = CompactorWorkload::default();
```
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/storage/src/hummock/compactor/mod.rs:385]: I haven't rebase main yet(so many conflicts), this has been removed in main branch.

File: src/storage/Cargo.toml (hunk 21 above)
[the remark this answers is not in the document]
Comment (hzxa21) [reply src/storage/Cargo.toml:70]: We tend to put all rpc client implementation in src/rpc_client. Can we put GrpcProxyClient there insated? To avoid confusion, we should call it GrpcCompactorProxyClient.

File: src/meta/src/hummock/compaction/mod.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (hzxa21) [reply src/meta/src/hummock/compaction/mod.rs:291]: It is not a bug but an optimization that needs to reverted in the new compaction strategy. I think it is better to address it in another PR for better visibility cc @Li0k @Little-Wallace

[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/meta/src/hummock/compaction/mod.rs:291]: Got it.

File: src/storage/Cargo.toml (hunk 21 above)
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/storage/Cargo.toml:70]: Previously storage crate only maintain all rpc client, but this pr introduced a rpc server, so we need to add tonic in storage crate.

File: src/storage/src/hummock/compactor/mod.rs (hunk 15 above)
Comment (wcy-fdu) [reply src/storage/src/hummock/compactor/mod.rs:717]: Will add retry strategy in next pr.

File: src/storage/src/hummock/sstable/sstable_object_id_manager.rs [hunk 24]
```diff
@@ -198,25 +199,57 @@ impl GetObjectId for Arc<SstableObjectIdManager> {
...
-        if let Some(first_element) = self.output_object_ids.pop_front() {
+        let mut guard = self.core.lock().await;
+        let core = guard.deref_mut();
+
+        if let Some(first_element) = core.output_object_ids.pop_front() {
             Ok(first_element)
         } else {
-            return Err(HummockError::other("Output object id runs out"));
+            tracing::warn!("The pre-allocated object ids are used up, and new object id are obtained through RPC.");
+            let request = GetNewSstIdsRequest { number: 10 };
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/sstable_object_id_manager.rs:240]: I think we can use this config https://github.com/risingwavelabs/risingwave/blob/94c59b8668f266b42593940942fbb64e0379bdb5/src/storage/src/opts.rs#L60 instead of hardcoding the fetch number here. This makes it easier to change if we find this number not reasonable in the future.

Comment (wcy-fdu) [reply src/storage/src/hummock/sstable/sstable_object_id_manager.rs]: Agree, fixed.

**Review by hzxa21 — COMMENTED** [review]
Rest LGTM. @Li0k PTAL

File: src/storage/hummock_test/src/compactor_tests.rs [hunk 25]
```diff
@@ -439,6 +429,56 @@ pub(crate) mod tests {
...
+            .unwrap()
+        {
+            // 3. compact
+            let (_tx, rx) = tokio::sync::oneshot::channel();
+            let (mut result_task, task_stats) = compact(
+                compact_ctx.clone(),
+                compact_task.clone(),
+                rx,
+                Box::new(sstable_object_id_manager.clone()),
+                filter_key_extractor_manager.clone(),
```
[the remark this answers is not in the document]
Comment (wcy-fdu) [reply src/storage/hummock_test/src/compactor_tests.rs:481]: It seems that it was added by mistake during conflict resolution, let me remove it.

File: src/storage/src/hummock/compactor/mod.rs [hunk 26]
```diff
@@ -603,3 +606,206 @@ pub fn start_compactor(
 
     (join_handle, shutdown_tx)
 }
+
+/// The background compaction thread that receives compaction tasks from hummock compaction
+/// manager and runs compaction tasks.
+#[cfg_attr(coverage, no_coverage)]
+#[allow(clippy::too_many_arguments)]
```
Comment (Li0k) [inline src/storage/src/hummock/compactor/mod.rs:613]: Is it need `too_many_arguments` ?

File: src/meta/src/rpc/service/hummock_service.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (Li0k) [reply src/meta/src/rpc/service/hummock_service.rs:529]: This confuses me, report_compaction_task is not part of the hummock service but proxy.

The way I understand the implementation,  it should be implemented using a client that is independent of the Proxy and then processed by the Proxy. Because we only need to handle a few interfaces.

Feel free to ignore me, we can add comments to it for context.

**Review by Li0k — APPROVED** [review]
Rest LGTM, thanks for the effort

**Comment by Li0k** [comment]
The pr has reached 100 commits, do we need to rebase / re-enable a pr ?

**Comment by wcy-fdu** [comment]
> The pr has reached 100 commits, do we need to rebase / re-enable a pr ?

A large part of the commits here come from the first three pre-order PRs.

File: src/storage/src/hummock/compactor/mod.rs [hunk 27]
```diff
@@ -608,3 +614,205 @@ pub fn start_compactor(
...
+                }
+
+            };
+            match request {
+                Some(request) => {
+                    let context = context.clone();
+                    let shutdown = shutdown_map.clone();
+
+                    let mut report_task_client = grpc_proxy_client.clone();
+                    let get_sst_id_client = grpc_proxy_client.clone();
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor/mod.rs:671]: The previous issue is not entirely resolved. We only need to clone once and don't need to hold two clients here.

**Review by hzxa21 — CHANGES_REQUESTED** [review]
Rest LGTM.

**Review by hzxa21 — APPROVED** [review]
Sorry. I mistakenly click "Request Change".
